homeassistant/components/keymitt_ble/__init__.py

```diff
@@ -0,0 +1,133 @@
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up this integration using UI."""
+    hass.data.setdefault(DOMAIN, {})
+    token: str = entry.data[CONF_ACCESS_TOKEN]
+    bdaddr: str = entry.data[CONF_BDADDR]
+    ble_device: BLEDevice | None = bluetooth.async_ble_device_from_address(
+        hass, bdaddr
+    )
+    if not ble_device:
+        raise ConfigEntryNotReady(f"Could not find MicroBot with address {bdaddr}")
+    client = MicroBotApiClient(
+        device=ble_device,
+        token=token,
+        retry_count=DEFAULT_RETRY_COUNT,
+    )
+    coordinator = MicroBotDataUpdateCoordinator(
+        hass, client=client, ble_device=ble_device
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(
+        entry, PLATFORMS
+    )
+
+    async def calibrate(call: ServiceCall) -> None:
+        _LOGGER.debug("Calibrate service called")
+        depth = call.data["depth"]
+        duration = call.data["duration"]
+        mode = call.data["mode"]
+        await coordinator.api.calibrate(depth, duration, mode)
+
+    hass.services.async_register(DOMAIN, "calibrate", calibrate)
+    return True
+
+
+class MicroBotDataUpdateCoordinator(PassiveBluetoothDataUpdateCoordinator):
+    """Class to manage fetching data from the MicroBot."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: MicroBotApiClient,
+        ble_device: BLEDevice,
+    ) -> None:
+        """Initialize."""
+        self.api = client
+        self.platforms: list[str] = []
+        self._ready_event = asyncio.Event()
+        self.data: dict[str, Any] = {}
+        self.ble_device = ble_device
+
+        super().__init__(
+            hass, _LOGGER, ble_device.address, bluetooth.BluetoothScanningMode.ACTIVE
+        )
+
+    @callback
+    def _async_handle_bluetooth_event(
+        self,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Handle a Bluetooth event."""
+        super()._async_handle_bluetooth_event(service_info, change)
+        if adv := parse_advertisement_data(
+            service_info.device, service_info.advertisement
+        ):
+            self.data = adv.data
+            if self.data:
+                self._ready_event.set()
+            _LOGGER.debug("%s: MicroBot data: %s", self.ble_device.address, self.data)
+            self.api.update_from_advertisement(adv)
+        self.async_update_listeners()
+
+    async def async_wait_ready(self) -> bool:
+        """Wait for the device to be ready."""
+        with contextlib.suppress(asyncio.TimeoutError):
+            async with async_timeout.timeout(55):
+                await self._ready_event.wait()
+                return True
+        return False
+
+    @property
+    def available(self) -> bool:
+        """Return true if the switch is available."""
+        return True
```

Comment:
Do we need to override the available property?  How does this handle the device disappearing from the Bluetooth stack?

Reply:
I don't fully understand how the coordinator works here so going to have to do some researching to try and figure it out